[custom-devices] add screenshotting support

ardera opened the pull request:
This implements #79678.

I tested it on my device and the flutter SDK part of this works. However it's still not possible to take screenshots of the app, at least when using flutter-pi because the device-part of screenshotting support is missing. I.e. there's currently no ready-to-use command to take screenshots of the Raspberry Pi when using KMS directly without X11 / Wayland. You can get a fbdev screenshot (and that's what I tested) and that works, but only shows the console. So in the future I'll probably add some kind of support inside flutter-pi to take screenshots.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

ardera committed:
add screenshotting support

jonahwilliams commented:
This would use a platform specific screenshot mechanism right?

Context is that I also have https://github.com/flutter/flutter/pull/80616 as a WIP which will allow falling back to a "rasterizer" screenshot. That should allow all platforms to have some level of screenshot support.

ardera commented:
> This would use a platform specific screenshot mechanism right?
>
> Context is that I also have #80616 as a WIP which will allow falling back to a "rasterizer" screenshot. That should allow all platforms to have some level of screenshot support.

Yep exactly. Just adds a new command to the config that will take the screenshot and store it in some path. If the command isn't present, the CustomDevice will report that screenshotting isn't supported.

jonahwilliams commented on packages/flutter_tools/lib/src/custom_devices/custom_device.dart:
```diff
@@ -594,6 +595,33 @@ class CustomDevice extends Device {
+    final RunResult result = await _processUtils.run(interpolated, throwOnError: true);
+
+    // remove all non-base64 characters.
+    final String trimmedStdout = result.stdout.trim().replaceAll(RegExp(r'[^a-zA-Z0-9+/]+'), '');
```
This seems like a bit of an awkward way of communicating the data. If there is non-base64 data output it could mean that the command failed unexpectedly, in which case you might just be creating an "image" out of a garbled string. Seems like failing would be the better choice

ardera replied:
My thought was that if something goes wrong, `ssh` should exit with an error code, that's why `throwOnError` is true and nothing is trying to catch the exception (don't really know what to do except that, should it just silently fail?)

Though, at least the `|` doesn't pass on failures, so maybe it'll run into the case you mentioned when the `cat` command unexpectedly fails. I'll change it

jonahwilliams commented on packages/flutter_tools/test/general.shard/custom_devices/custom_device_test.dart:
```diff
@@ -519,6 +520,63 @@ void main() {
+
+    await device.takeScreenshot(screenshotFile);
+    expect(screenshotCommandWasExecuted, true);
+    expect(screenshotFile.existsSync(), true);
```
nit: you can use `exists` from package:file_testing

ardera replied:
done

jonahwilliams reviewed (commented):
Overall LGTM, except for comment on the regex replace of output

ardera committed:
- use exists matcher for testing instead of matching existsSync against true
- don't trim the screenshot output in the SDK, let the command do it
- change the default screenshot command to use 'tr' to remove whitespaces, tabs and newlines
- remove some logging in takeScreenshot since processUtils.run will do the necessary logging itself

ardera commented:
how am I getting this allowlist error? The problem with it started happening way after the base master commit I'm using and the bundle builder refactor (which used the same upstream commit as base) tested fine

jonahwilliams commented:
There was a bug in the test script, you'll need to pull lastest master to get the fix

jonahwilliams commented:
@ardera do you need any help getting synced to a good flutter hash?

ardera commented:
> @ardera do you need any help getting synced to a good flutter hash?

Thanks but I just didn't have time on sunday and wanted to do it later instead 😄 probably one reason why you guys don't give time estimates ;)

jonahwilliams commented:
Fair enough 😄

ardera commented:
So there's one test failing, but I'm not sure it's related. Seems like it times out after 25s, but some other tests take long as well (the `Test is taking a long time` message is printed out a few other times)
```
04:38 +9 ~14: test\integration.shard\overall_experience_test.dart: flutter error messages include a DevTools link                                                                                      
Test is taking a long time.
Expected state transitions:
    0 NOW WAITING FOR> /^An Observatory debugger and profiler on Flutter test device is available at: /
    1                  /^The Flutter DevTools debugger and profiler on Flutter test device is available at: /
    2                  /^Performing hot reload\.\.\./
    3                  /^Reloaded 0 libraries in [0-9]+ms./
So far nothing has been logged; use debug:true to print all output.
(streaming all logs from this point on...)
stdout: Syncing files to device Flutter test device...                     265ms
stdout: 
stdout: Flutter run key commands.
stdout: r Hot reload. 
stdout: R Hot restart.
stdout: h List all available interactive commands.
stdout: d Detach (terminate "flutter run" but leave application running).
stdout: c Clear the screen
stdout: q Quit (terminate the application on the device).
stdout: 
stdout:  Running with sound null safety 
stdout: 
stdout: An Observatory debugger and profiler on Flutter test device is available at: http://127.0.0.1:53624/49ye2qa6uGk=/
(matched /^An Observatory debugger and profiler on Flutter test device is available at: /)
(process is not quitting, trying to send a "q" just in case that helps)
(a functional test should never reach this point)
stdout: 
stdout: Application finished.
(process terminated with exit code 0)
The subprocess terminated before all the expected transitions had been matched.
The transition that we were hoping to see next but that we never saw was:
    1 NOW WAITING FOR> /^The Flutter DevTools debugger and profiler on Flutter test device is available at: /

04:50 +9 ~14 -1: test\integration.shard\overall_experience_test.dart: flutter error messages include a DevTools link [E]                                                                               
  Missed some expected transitions.
  test\integration.shard\overall_experience_test.dart 275:5  runFlutter
```

jonahwilliams commented:
That is unfortunately another broken test, we just landed https://github.com/flutter/flutter/commit/1b4f4482b75b92a95807ee4337798227c439dc61 to work around it

ardera commented:
Okay, rebased onto that now

jonahwilliams reviewed (approved):
LGTM
